File: solutions/clojure/006-reverse-integer.clj
;; Problem 6: Reverse Integer (LeetCode 7)
;; Difficulty: Med
;; Language: Clojure
;; 
(defn reverse-integer [x]
  (let [sign (if (neg? x) -1 1)
        reversed (->> (Math/abs (long x))
                      str
                      clojure.string/reverse
                      Long/parseLong
                      (* sign))
        max-int (dec (long (Math/pow 2 31)))
        min-int (- (long (Math/pow 2 31)))]
    (if (or (> reversed max-int) (< reversed min-int)) 0 reversed)))

File: solutions/rust/006-reverse-integer.rs
/**
 * Problem 6: Reverse Integer (LeetCode 7)
 * Difficulty: Med
 * Language: Rust
 */
pub fn reverse(x: i32) -> i32 {
    let sign = if x < 0 { -1i64 } else { 1i64 };
    let reversed = x.abs()
        .to_string()
        .chars()
        .rev()
        .collect::<String>()
        .parse::<i64>()
        .unwrap_or(0)
        * sign;
    if reversed > i32::MAX as i64 || reversed < i32::MIN as i64 {
        0
    } else {
        reversed as i32
    }
}
